### ISSUE TITLE
Show session expiration modal on 401/403  errors in looker heartbeat call

### ISSUE THREAD

**Issue by vikasrohit** [appirio-tech/connect-app#3745]
Currently we have a session timer which triggers some time (2 minutes right now) before the session expiration and shows the modal to the user to refresh the reports. However, we are seeing many incidents where looker is throwing 401/403 some how e.g. first GET request of that embed URL is cancelled (network is slow or any other error in the page prevented that call to complete) which makes the URL invalid for second invocation OR user login and logout into full topcoder session in separate tab.
Purpose of this issue is to do investigation about solution for this problem.

**Comment by vikasrohit** [comment]
fyi @maxceem

**Comment by maxceem** [comment]
@vikasrohit if I understand it correctly, this has to be fixed by 48fb21e4e15f8561e820dc289f36306bddb64675. Let me know if you still experience this issue.

**Comment by vikasrohit** [comment]
@maxceem this is slightly different than what you fixed in that commit. I hoping for handling the error event when looker's internal health check (`/hearbeat`) fails with 401 0r 403. If can handle that event, we can show the modal in that case as well without waiting for timer to reach.

**Comment by maxceem** [comment]
@vikasrohit do I understand that your plan is like this?
- in Project Service raise an event to Kafka
- in TC Notification Service process the Kafka event and create a notification
- in Connect App, we are checking for notifications every minute, so as soon as we've got a notification that Looker internally failed we are showing the modal

**Comment by vikasrohit** [comment]
No, @maxceem I think I was not clear from starting and I think because you are not able to see any report, you are not seeing the hearbeat calls that looker is making behind the scenes.
When looker actually renders a report, it keeps sending heartbeat api calls to its server to check how much session time is left. And this call returns with 401/403 when due to any reason the session is expired at looker server. I want to catch this event and show the modal.

**Comment by maxceem** [comment]
Do I understand it correctly, that this `heartbeat` request is triggered using some code inside the `iframe`? So all this logic happens inside `iframe`?

If so I guess there is no common solution, as we cannot access the `iframe` on another domain as per browsers security policies, unless we can implement some logic inside `iframe` to talk to the "host".

Or maybe there is some kind of API of Looker? I guess we would need to have more details on how it works to find a solution.

**Comment by vikasrohit** [comment]
> Do I understand it correctly, that this heartbeat request is triggered using some code inside the iframe? So all this logic happens inside iframe?

Yes.

> If so I guess there is no common solution, as we cannot access the iframe on another domain as per browsers security policies, unless we can implement some logic inside iframe to talk to the "host".

Understood. I was thinking of some event handler on iframe itself like we have onload, if we have onerror, we can put some logic there. But if it is not, it is not, we can not do anything.

**Comment by maxceem** [comment]
I guess to check it out we would need to have a demo on the DEV. So we can play with a real example and see if there is any way to find out about this error in iframe from the "host".

**Comment by vikasrohit** [comment]
Okay, lets leave it for now, I will setup something in dev for reports after publicly releasing the reports feature. we can take it afterwards.
